Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -0,0 +1,237 @@
+ * `DivSequence`: a sequence indexed by integers is a divisibility sequence.
+ * `EllDivSequence`: a sequence indexed by integers is an EDS.
+ * `normEDS'`: the canonical example of a normalised EDS indexed by `ℕ`.
+ * `normEDS`: the canonical example of a normalised EDS indexed by `ℤ`.
+
+## Main statements
+
+ * TODO: prove that `normEDS` is an `EllDivSequence`.
+ * TODO: prove that a general normalised `EllDivSequence` can be given by `normEDS`.
+
+## Implementation notes
+
+`EllDivSequence' b c d n` is defined in terms of the private definition `EllDivSequence'' b c d n`,
+which are equal when `n` is odd and differ by a factor of `b` when `n` is even. This coincides with
+the reference since both agree for `EllDivSequence' b c d 2` and for `EllDivSequence' b c d 4`, and
+the correct factors of `b` are removed in `EllDivSequence' b c d (2 * (m + 2) + 1)` and in
+`EllDivSequence' b c d (2 * (m + 3))`. This is done to avoid the necessity for ring division by `b`
+in the inductive definition of `EllDivSequence' b c d (2 * (m + 3))`. The idea is that, an easy
+lemma shows that `EllDivSequence' b c d (2 * (m + 3))` always contains a factor of `b`, so it is
+possible to remove a factor of `b` a posteriori, but stating this lemma requires first defining
+`EllDivSequence' b c d (2 * (m + 3))`, which requires having this factor of `b` a priori.
+
+## References
+
+M Ward, *Memoir on Elliptic Divisibility Sequences*
+
+## Tags
+
+elliptic, divisibility, sequence
+-/
+
+universe u v w
+
+variable {R : Type u} [CommRing R]
+
+/-- The proposition that a sequence indexed by integers is an elliptic sequence. -/
+def EllSequence (W : ℤ → R) : Prop :=
+  ∀ m n r : ℤ, W (m + n) * W (m - n) * W r ^ 2 =
+    W (m + r) * W (m - r) * W n ^ 2 - W (n + r) * W (n - r) * W m ^ 2
+
+/-- The proposition that a sequence indexed by integers is a divisibility sequence. -/
+def DivSequence (W : ℤ → R) : Prop :=
+  ∀ m n : ℕ, m ∣ n → W m ∣ W n
+
+/-- The proposition that a sequence indexed by integers is an EDS. -/
+def EllDivSequence (W : ℤ → R) : Prop :=
+  EllSequence W ∧ DivSequence W
+
+lemma ellSequence_id : EllSequence id :=
+  fun _ _ _ => by simp only [id_eq]; ring1
+
+lemma divSequence_id : DivSequence id :=
+  fun _ _ => Int.ofNat_dvd.mpr
+
+/-- The identity sequence is an EDS. -/
+theorem ellDivSequence_id : EllDivSequence id :=
+  ⟨ellSequence_id, divSequence_id⟩
+
+lemma ellSequence_mul (x : R) {W : ℤ → R} (h : EllSequence W) : EllSequence (x • W) := fun m n r =>
+  by linear_combination (norm := (simp only [Pi.smul_apply, smul_eq_mul]; ring1)) x ^ 4 * h m n r
+
+lemma divSequence_mul (x : R) {W : ℤ → R} (h : DivSequence W) : DivSequence (x • W) :=
+  fun m n r => mul_dvd_mul_left x <| h m n r
+
+lemma ellDivSequence_mul (x : R) {W : ℤ → R} (h : EllDivSequence W) : EllDivSequence (x • W) :=
+  ⟨ellSequence_mul x h.left, divSequence_mul x h.right⟩
+
+private def normEDS'' (b c d : R) : ℕ → R
```

Comment:
Do you prefer the current names `normEDS''`/`normEDS'` or what I suggested (`normEDSAux`/`natNormEDS`)?

Reply:
I prefer my current one - I want to try to keep the capitalisation approximately the same for all three

Reply:
I think `normEDSAux` doesn't interfere with capitalization?
Actually, how about `Nat.normEDSAux`, `Nat.normEDS` and `Int.normEDS` which would give access to dot notation. Plus it's a bit weird to have these in the root namespace.

Reply:
I thought of doing `Nat.normEDS` as well originally, but I don't like `Int.normEDS` (since EDSs are already supposed to be indexed by `Int` by my convention - I might change this in a future PR and update this accordingly). I also don't want to these to lie in the `Nat`/`Int` namespace and allow things like `n.normEDS b c d` because the `n` should really come after (to actually look like it's indexed by `n`). I don't think this will interfere much with the rest of mathlib despite being in the root namespace. If I were to keep `normEDS'` and `normEDS`, then `normEDSAux` should really be `normEDS'Aux` (but that's a bit weird), and it's a private definition anyway so the name isn't too important.

Reply:
Okay, I'm fine with the names now. If it looks okay to the maintainer @alexjbest please merge.

Reply:
maintainer merge